### PR TITLE
Alternative batching strategy

### PR DESCRIPTION
**Pull request by andimarek**
Alternative batching strategy which allows for more control than the default one.

### CODE REVIEW

File: src/main/java/graphql/execution/instrumentation/dataloader/ChainedDataLoader.java [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+
+import graphql.ExperimentalApi;
+
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionStage;
+import java.util.function.Function;
+
+
+@ExperimentalApi
+public final class ChainedDataLoader<T, U> extends CompletableFuture<U> {
```
Comment (bbakerman) [inline src/main/java/graphql/execution/instrumentation/dataloader/ChainedDataLoader.java:11]: i feel like this is the wrong package for this this class

Its nothing to do with `instrumentation`

I suggest a new one `graphql.dataloader`

File: src/main/java/graphql/execution/instrumentation/dataloader/ChainedDataLoader.java [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+    }
+
+    public CompletableFuture<Void> getSecondDataLoaderCalled() {
+        return secondDataLoaderCalled;
+    }
+
+    @ExperimentalApi
+    public static <T, U> ChainedDataLoader<T, U> two(CompletableFuture<T> firstDataLoaderResult, Function<? super T, ? extends CompletableFuture<U>> applyFunction) {
+        return new ChainedDataLoader<>(firstDataLoaderResult, applyFunction);
+    }
```
Comment (bbakerman) [inline src/main/java/graphql/execution/instrumentation/dataloader/ChainedDataLoader.java:56]: what about `three` and `four` DLs say

File: src/main/java/graphql/execution/instrumentation/dataloader/BatchLoadingDispatchStrategy.java [hunk 3]
```diff
@@ -0,0 +1,301 @@
...
+        }
+        return resultPath.getLevel() - trivialCount;
+    }
+
+
+    @Override
+    public void fieldFetched(ExecutionContext executionContext, ExecutionStrategyParameters parameters, DataFetcher<?> dataFetcher, CompletableFuture<Object> fetchedValue) {
+        ResultPath path = parameters.getPath();
+        int level = getLevelForPath(path);
+        System.out.println("field fetched at " + path);
```
Comment (bbakerman) [inline src/main/java/graphql/execution/instrumentation/dataloader/BatchLoadingDispatchStrategy.java:138]: sout ?

File: src/main/java/graphql/execution/Execution.java [hunk 4]
```diff
@@ -223,6 +224,10 @@ private CompletableFuture<ExecutionResult> incrementalSupport(ExecutionContext e
     }
 
     private DataLoaderDispatchStrategy createDataLoaderDispatchStrategy(ExecutionContext executionContext, ExecutionStrategy executionStrategy) {
+        if (executionContext.getGraphQLContext().get(DataLoaderDispatchStrategy.CUSTOM_STRATEGY_KEY) != null) {
+            Function<ExecutionContext, DataLoaderDispatchStrategy> customStrategy = executionContext.getGraphQLContext().get(DataLoaderDispatchStrategy.CUSTOM_STRATEGY_KEY);
+            return customStrategy.apply(executionContext);
+        }
```
Comment (bbakerman) [inline src/main/java/graphql/execution/Execution.java:230]: We have

```public @interface ExperimentalApi {
    /**
     * The key that should be associated with a boolean value which indicates whether @defer and @stream behaviour is enabled for this execution.
     */
    String ENABLE_INCREMENTAL_SUPPORT  = "ENABLE_INCREMENTAL_SUPPORT";
}
```

I think we should continue using that place for experimental stuff

**Comment by andimarek** [comment]
Superseded by https://github.com/graphql-java/graphql-java/pull

**Pull request closed without merging**